config/wolvm.man
				      configure_wolvm.py

NAME
	configure_wolvm.py - python script generating configuration file for wolvm service

SYNOPSIS
	wolvm.py configure 	[-s SOURCE_CONFIG_FILE] [-d KERNEL_CONFIG_FILE]
	wolvm.py start 		[-p PATH_TO_KERNEL]
	wolvm.py stop 		[-p PATH_TO_KERNEL]
	wolvm.py (-h | --help | help)
	
DESCRIPTION
	configure_wolvm.py will generate the appropriate configuration file for wolvm. It will try 
	to get necessary information from every vm manager installed on host machine. For a vm to 
	awaken by wolvm, its vm manager must be appropriately registered in SOURCE_CONFIG_FILE (for
	explanations see the "FILES" section of this document).

OPTIONS 
	configure
		generate wolvm config file "KERNEL_CONFIG_FILE" (default is vm.list).
		configure_wolvm.py will use "SOURCE_CONFIG_FILE" to determine how to get the list
		of vms available and their associated MAC address.

	start/stop 
		will respectively start/stop the kernel located at PATH_TO_KERNEL if you specify it,
		or at the location <kernelPath> defined in SOURCE_CONFIG_FILE.

	-h, --help, help
		print help

FILES
	SOURCE_CONFIG_FILE
		This file contains information used by configure_wolvm.py to communicate with VM 
		managers.
		We strongly advise you to use DTD provided to check if you made no mistakes editing
		this file.
		If you want to add a non supported VM manager you must enter the appropriate 
		information in this file as specified below:
			SOURCE_CONFIG_FILE is an xml file witch structure is:
			<WolVM>
				<kernelPath>pathToKernel</kernelPath>
				<VMTypes>
					<VMType id='id' name='name' version='version'>
						<check>VMmanagerCommand</check>
						<list regex numRegexGroup>commandList</list>
						<mac  regex numRegexGroup>commandMac</mac>
						<launchVM>commandLaunch</launchVM>
					</VMType>
				</VMtypes>
				<VMmanagers>
					<VMmanager type='type' name="VMmanagerName" path="path"></VMmanager>
				</VMmanagers>
			</WolVM>
			The <VMTypes> field contains the definition of a VMmanager type. However
			this field contains information related to the VMmanager in itself. On
			the other hand <VMmanagers> contains information related to the way the
			VMmanager is installedon the host system.
			For a VM manager to be used by configure_wolvm.py it needs to be configured
			both in <VMTypes> and in <VMmanagers>.

			To add a VM manager you must add a new <VMType> field where:
				-id		is an unique key designing the VM manager
				-name: 		is the name of the VM manager

				-pathToVMmanager: is the name of the VM manager's command line program
				-<list>
					-commandList:	command to use to get the list of VMs managed by
							the VM manager
					-regex:		regular expression to use retrieve the name (or ID)
							of each VM (and only its name or ID).
							This regular expression will be applied to the
							output of the "list" command.
					-numRegexGroup:	number of the group matched by "regexList" that
							cointains the VM name or ID (and only the name or
							id)
				-<mac>
					-commandMac	command to use to get the MAC address associated
							to the VM 'ID'. The ID of every mac found with 
							"list" will be appended to the end of this command.
					-regex:		regular expression to use to retrieve the MAC 
							address from the output of "mac".
					-numRegexGroup:	number of the group matched by "regexMAC" that
							cointains the VM MAC
				-commandLaunch: command to use to launch a VM managed by the VM
						manager. VM name or ID will be appended at the end
						of this command.


			To make configure_wolvm.py use a vm manager which type has been defined in
			<VMTypes> you need to create a <VMmanager> field in <VMmanagers> containing
			the information specified below:
				-type:		reference to the id of the corresponding <VMType> field
				-name:		the name you want to give to this VM manager

				-path: 		is the absolute path to the FOLDER containing 
						your VM manager binary (last caracter must be 
						'/' for UNIX and '\' for windows because the name
						of the executable (in <check>) will be appended at 
						the end of this string


			you must also specify the absolute path to the kernel executable so that you can 
			launch it using configure_wolvm.py
		
		WOLVM_PARSE_FOLDER is a special option you can specify in both the <list> and <mac>
		field. It must be followed by an absolute path to a folder (ending with / for UNIX 
		or \\ for windows). If this option is specified in both <list> and <mac> fields, the
		system will then look for mac addresses in every file which name matches the <list>
		regex. The content each file will be parsed by trying to match every line with the
		mac regex. The first positive match ends the parsing of the file and starts the
		analisys of the next one. 
		N.B: Commented lines in every file can match if you don't give a regex accurate 
		enough.

	KERNEL_CONFIG_FILE
		This file is the file in which is stored the association between the MAC addresses and the
		used to launch them. The MAC address and the launching command of one VM are on the same 
		line, separated by a coma and *no space*.
		This file can be manually edited but those modifications will be lost the next time 
		configure_wolvm.py is being run.

		THIS FILE MUST IN THE SAME FOLDER THAN THE KERNEL AND MUST BE NAMED 'vm.list' (to change 
		that you will have to recompile the kernel)

SEE ALSO
	Regular expressions in python: http://docs.python.org/library/re.html

AUTHORS




daemon/wolvm.c
/* WOLVM, a Wake On Lan extension for virtual machines.
 * Below is the WOLVM kernel source code.
 *
 * File		: wolvm_windows.c
 *
 * Description	: Source code of the WLOVM kernel.
 * 		  The kernel's job is to detect the magic packets and to wake
 * 		  up the Virtual Machines which MAC address is in the magick packet.
 * 		  Of course this is if there is a VM with the corresponding MAC on the host 
 * 		  system.
 *
 * Last Modif	: 06/24/2012
 *
 * Authors	: MERLINI Adrien - <adrien.merlini@telecom-bretagne.eu>
 * 		  ZHANG Chi 	 - <chi.zhang@telecom-bretagne.eu>
 *
 * Contact	: BEGLIN   Marianne - <marianne.beglin@telecom-bretagne.eu>
 *                MERLINI  Adrien   - <adrien.merlini@telecom-bretagne.eu>
 *                NGUYEN   Olivier  - <olivier.nguyen@telecom-bretagne.eu>
 *                Zhang    Chi      - <chi.zhang@telecom-bretagne.eu>
 *
 * Requires	: You must have installed WinPcap (drivers + DLLs) - http://www.winpcap.org/install/default.htm
 * 		  To compile this code you also need to download and set the Winpcap Developer Resources - http://www.winpcap.org/devel.htm
 *
 * Compile	: Use your favorite IDE, but don't forget to link the compiler to the WinPcap libs
 * 		  and to place the Winpcap headers where your compiler needs you to. 
 * */


#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <pcap.h>
#include <windows.h>

#undef ERROR
#include "wolvm_config.h"

/* Path to the log file */
#define LOG_FILE        "C:\\Windows\\Temp\\WOLVM.log"
/* Path to the KERNEL_CONFIG_FILE, this file is refered to as vm.list in the WOLVM 
 * configuring script*/
#define KERNEL_CONFIG_FILE      "vm.list"
/* Max size of log messages */
#define LOG_MAX_SIZE    100
/* Max size of the Mac addresses returned by pcap. 16 should be enough but... */
#define MAC_LEN 20
/* Max size of the commands extracted from vm.list */
#define CMD_LEN 100
/* Max number of the packets bites to be read */
#define MAXBYTES2CAPTURE 2048

/* Seems to be defined in windows.h, undefining it to */
#undef ERROR


/*--------------------------------------------------------------------------------------------------*/
/*----------------------------------------------HEADERS---------------------------------------------*/
/*--------------------------------------------------------------------------------------------------*/

/* Windows' service related varibles and functions */
SERVICE_STATUS ServiceStatus;
SERVICE_STATUS_HANDLE hStatus;

void ServiceMain(int argc, char** argv);
void ControlHandler(DWORD request);
int  InitService();

/* WOLVM functions */
int  log_message(char *message, LOGLVL logLevel);
void formatMac(char *mac);
void launchVM(char *mac);
void processPacket(const struct pcap_pkthdr* pkthdr, const u_char * packet);


/*--------------------------------------------------------------------------------------------------*/
/*-----------------------------------------GLOBAL VARIABLES-----------------------------------------*/
/*--------------------------------------------------------------------------------------------------*/

int verbosity;


/*--------------------------------------------------------------------------------------------------*/
/*----------------------------------------------SOURCES---------------------------------------------*/
/*--------------------------------------------------------------------------------------------------*/

int main()
{
	/* Main just connect the program to the Windows Service Manager */
	SERVICE_TABLE_ENTRY ServiceTable[2];

	ServiceTable[0].lpServiceName = "WOLVM";
	ServiceTable[0].lpServiceProc = (LPSERVICE_MAIN_FUNCTION)ServiceMain;

	ServiceTable[1].lpServiceName = NULL;
	ServiceTable[1].lpServiceProc = NULL;
	// Start the control dispatcher thread for our service
	if(!StartServiceCtrlDispatcher(ServiceTable))
	{
		/* If you see this message appear in your log file, you probably attempting to 
		 * launch the program directly.
		 * You must use the Service Manager to do so. 
		 * See README.txt.
		 * */
		log_message("The system is preveting service WOLVM from starting, see service.msc", FATAL);
		return EXIT_FAILURE;
	}

	return EXIT_SUCCESS;
}


/*--------------------------------------------------------------------------------------------------*/

void ServiceMain(int argc, char** argv)
{
	/* This function would be the 'real' main() if there were no need to communicate 
	 * with the service manager.
	 * */


	/* Windows Service specific code */
	int error;
	char *progName=argv[0];

	pcap_t *descr = NULL;
	char errbuf[PCAP_ERRBUF_SIZE], *device=NULL;
	char message[50];
	memset(errbuf,0,PCAP_ERRBUF_SIZE);
	struct pcap_pkthdr pkt_header;
	struct pcap_pkthdr * ppkt_header = &pkt_header;
	const u_char *pkt_data;


	ServiceStatus.dwServiceType        = SERVICE_WIN32;
	ServiceStatus.dwCurrentState       = SERVICE_START_PENDING;
	ServiceStatus.dwControlsAccepted   = SERVICE_ACCEPT_STOP | SERVICE_ACCEPT_SHUTDOWN;
	ServiceStatus.dwWin32ExitCode      = 0;
	ServiceStatus.dwServiceSpecificExitCode = 0;
	ServiceStatus.dwCheckPoint         = 0;
	ServiceStatus.dwWaitHint           = 0;

	hStatus = RegisterServiceCtrlHandler(
		"WOLVM",
		(LPHANDLER_FUNCTION)ControlHandler);
	if (hStatus == (SERVICE_STATUS_HANDLE)0)
	{
		// Registering Control Handler failed
		log_message("Registering control handler failed, aborting.", FATAL);
		return;
	}
	// Initialize Service
	error = InitService();
	if (error)
	{
		// Initialization failed
		log_message("Service initialisation failed, aborting.", FATAL);
		ServiceStatus.dwCurrentState       = SERVICE_STOPPED;
		ServiceStatus.dwWin32ExitCode      = -1;
		SetServiceStatus(hStatus, &ServiceStatus);
		return;
	}
	// We report the running status to SCM.
	ServiceStatus.dwCurrentState = SERVICE_RUNNING;
	SetServiceStatus (hStatus, &ServiceStatus);
	/* END OF Windows Service specific code */

	/* Loading config file information */
	readConf(opts);

	/* Parsing arguments (k & R method) */
	if(argc == 2)
	{
		if(strcmp(*++argv,"-v") == 0){
			setOption("verbosity", "1", opts);
		}else{
			setOption("iface", *argv, opts);
		}
	}

	else if(argc == 3){
		if(strcmp(*++argv,"-v") == 0){
			setOption("verbosity", "1", opts);
		}else{
			badUsage(progName);	
		}
		setOption("iface", *++argv, opts);
	}

	/* Starting proper work with correct options */
	verbosity = atoi(getOptValue("verbosity", opts));

	device = getOptValue("iface", opts);

	
	/* pcap starts working with the network */
	if(strncmp(device, "auto", 5) == 0)		 /* Get the name of the first device suitable for capture */
	{
		if ((device = pcap_lookupdev(errbuf)) == NULL)
		{
			log_message(errbuf, FATAL);
			exit(EXIT_FAILURE);
		}
	}

	sprintf(message, "Opening device %s", device);
	log_message(message, INFORMATION);

	/* Open device in promiscuous mode */
	if ((descr = pcap_open_live(device, MAXBYTES2CAPTURE, 1,  512, errbuf)) == NULL)
	{
		log_message(errbuf, FATAL);
		exit(EXIT_FAILURE);
	}

	// The worker loop of a service
	while (ServiceStatus.dwCurrentState == SERVICE_RUNNING)
	{
		if(pcap_next_ex(descr, &ppkt_header, &pkt_data)!=-1){
			processPacket(ppkt_header, pkt_data);
		}   
	}


	return;
}


/*--------------------------------------------------------------------------------------------------*/

void badUsage(char *progName)
{
	char message[100];
	sprintf(message, "Usage: %s [-v] [iface]\n", progName);
	log_message(message, FATAL);
	exit(EXIT_FAILURE);
}


/*--------------------------------------------------------------------------------------------------*/


// Service initialization
int InitService()
{
	return log_message("Monitoring started.", INFORMATION);
}


/*--------------------------------------------------------------------------------------------------*/

// Control handler function
void ControlHandler(DWORD request)
{
	switch(request)
	{
		case SERVICE_CONTROL_STOP:
			log_message("Monitoring stopped.", INFORMATION);

			ServiceStatus.dwWin32ExitCode = 0;
			ServiceStatus.dwCurrentState  = SERVICE_STOPPED;
			SetServiceStatus (hStatus, &ServiceStatus);
			return;

		case SERVICE_CONTROL_SHUTDOWN:
			log_message("Monitoring stopped.", INFORMATION);

			ServiceStatus.dwWin32ExitCode = 0;
			ServiceStatus.dwCurrentState  = SERVICE_STOPPED;
			SetServiceStatus (hStatus, &ServiceStatus);
			return;

		default:
			break;
	}

	// Report current status
	SetServiceStatus (hStatus,  &ServiceStatus);

	return;
}


/*--------------------------------------------------------------------------------------------------*/

int log_message(char *message, LOGLVL logLevel)
{
	/* Logging function with different loglevels */
	FILE *logfile	= NULL;
	time_t now	= time(NULL);
	char logMsg[LOG_MAX_SIZE]={'\0'};

	strcat(logMsg, ctime(&now));
	/* Suppressing the '\n' in the string returned by cdate */
	logMsg[24] = ' ';

	switch(logLevel)
	{
		case DEBUG:
			if(!verbosity) return 0;
			strcat(logMsg, "(DD) ");
			strncat(logMsg, message, LOG_MAX_SIZE - 6);
			break;
		case INFORMATION:
			strcat(logMsg, "(II) ");
			strncat(logMsg, message, LOG_MAX_SIZE - 6);
			break;
		case WARNING:
			strcat(logMsg, "(WW) ");
			strncat(logMsg, message, LOG_MAX_SIZE - 6);
			break;
		case ERROR:
			strcat(logMsg, "(EE) ");
			strncat(logMsg, message, LOG_MAX_SIZE - 6);
			break;
		case FATAL:
			strcat(logMsg, "(FF) ");
			strncat(logMsg, message, LOG_MAX_SIZE - 6);
			break;
		default:
			break;
	}

	logfile=fopen(LOG_FILE, "a");
	if(!logfile) return -1;
	fprintf(logfile,"%s\n",logMsg);
	fclose(logfile);

	return 0;
}


/*--------------------------------------------------------------------------------------------------*/

// processPacket(): Callback function called by pcap_loop() everytime a packet
// arrives to the network card.
// This function determines wether a packet is a magic packet or not.
// If it is, launches the appropriate treatment.

void processPacket(const struct pcap_pkthdr* pkthdr, const u_char * packet)
{

	unsigned int i = 0, backUpI = 0, j, k, pktLen = pkthdr->len;
	int syncro, consecutif;
	u_char mac[6];
	char macToLog[50], macToLaunch[17];

	while(backUpI + 96 <= pktLen)
	{
		syncro = 0;
		consecutif = 1;
		i = backUpI - 1;
		while(++i<pktLen && syncro<6)
		{
			if(packet[i] == 0xff && consecutif)
			{
				syncro++;
			}
			else if(packet[i] == 0xff && !consecutif)
			{
				syncro = 1;
				consecutif = 1;
			}
			else if(consecutif)
			{
				consecutif = 0;
			}
		}
		backUpI=i;
		if(syncro == 6 && i + 96 == pktLen)
		{
			for(j = 0; j < 6; j++)
			{
				mac[j] = packet[i + j];
			}
			i += 6;
			j = -1;
			consecutif = 1;
			while(++j < 15 && consecutif)
			{
				for(k = 0; k < 6; k++)
				{
					if(packet[i + k] != mac[k])
						consecutif = 0;
				}
				i += 6;
			}

			/* If the algo has recognized a magic packet */
			if(consecutif)
			{
				sprintf(macToLaunch,"%2x%2x%2x%2x%2x%2x", mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
				launchVM(macToLaunch);
				sprintf(macToLog,"Received Magic Paquet destined to: %x:%x:%x:%x:%x:%x", mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
				log_message(macToLog, DEBUG);
			}
		}
	}
	return;
}


/*--------------------------------------------------------------------------------------------------*/

void launchVM(char *mac){
	/* function called when a magick packet is received. It parses vm.list 
	 * and if there is a matching mac ask the system to launch the corresponding VM.
	 * */

	FILE *vmlist;
	char temp[CMD_LEN+MAC_LEN];
	char *tempToken;

	formatMac(mac);
	/*Opening KERNEL_CONFIG_FILE*/
	if(!(vmlist = fopen(KERNEL_CONFIG_FILE,"r"))){
		log_message("Fail to find the KERNEL_CONFIG_FILE.\nExiting.", FATAL);
		exit(EXIT_FAILURE);
	}

	while(fgets(temp,CMD_LEN+MAC_LEN,vmlist)){
		tempToken=strtok(temp, ",");
		if(!strncmp(tempToken, mac, MAC_LEN)){
			tempToken = strtok(NULL, ",");
			tempToken ? system(tempToken) : log_message("KERNEL_CONFIG_FILE contains a non well-formated line", WARNING);
			break;
		}
	}
	fclose(vmlist);
}


/*--------------------------------------------------------------------------------------------------*/

/* Function used to make sure that the mac has no ' '. This format MUST be coherent with the format
 * used in vm.list.
 * */
void formatMac(char *mac)
{
	while(*mac){
		if(*mac == ' '){
			*mac = '0';
		}
		++mac;
	}
}

	
/*--------------------------------------------------------------------------------------------------*/

/* EOF */
